### PR TITLE
Add a Contributor's Guide feature to the docs

### PR DESCRIPTION
**Pull request by gspetro-NOAA**
## DESCRIPTION OF CHANGES: 
Added a Contributor's Guide Chapter so that contributors to the SRW App understand the Issue creation & PR Process, required code & documentation standards, etc. 
NOTE: Reviewers only need to review the Contributor's Guide chapter. Changes to all other chapters will be accounted for in PR #212 , which should be merged first. 

## TESTS CONDUCTED: 
None required. Text only changes. Built on ReadTheDocs to make sure there are no problems. 

## DEPENDENCIES:
PR #212 needs to be merged first, since this contains all of the same changes PLUS the contributor's guide. 

## DOCUMENTATION:
Here is the link to the [RTD documentation](https://srw-contributors-guide.readthedocs.io/en/latest/ContributorsGuide.html) for the Contributor's Guide.

## ISSUE: 
Related to Issue #211 

## CONTRIBUTORS (optional): 
@mark-a-potts @mkavulich

### CODE REVIEW

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: I think the addition of a textonly/[name] branch convention is useful.  I'll let others comment, but if others feel the same, I think you can remove the comment here.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: Is the Development section commented out?

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: You might want to briefly mention the PR template that will auto-fill when a PR is first opened.  It's described in detail below, but it might be good to say it exists here.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: "... delete the branch on their fork..."

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst:149]: Looks like we need to add a new issue template for text-only changes.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst:158]: Since there are more types of branches than just "feature" or "bug-fix", I guess we should make this more general?

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: SRW -> SRW App, PR's -> PRs

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: dependent branches -> dependent branches in the developer's fork

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: The code developer should update the branches/hashes in their ufs-srweather-app repo once their corresponding regional_workflow PR has been merged.  A developer should mention in their ufs-srweather-app repo PR description that they are temporarily pointing to a branch/hash in their fork of regional_worklfow, but that it will be updated once the corresponding regional_workflow PR is merged.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: Urgency can also be signaled through labels that can be attached to the PR, such as "high priority".  You can see examples of these in the closed PRs.  They're added on the right hand side of a PR once they're opened.  We should probably encourage developers to add a label for the type of PR ("bug", "feature", "enhancement", etc.) and the priority level.

EDIT: I see you have a label section below.  Maybe expanding that section with this information would make sense.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: Can you expand on this comment?  Was this related to the wiki?

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: This can be done by requesting specific reviewers (above and beyond the normal set of code owners) on the right-hand side of the PR.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst:248]: Since the repo is public, anyone should be able to open a PR.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst:265]: "Externals should point...."  .. with the exception of a PR into ufs-srweather-app that is referencing code in a contributors fork for a regional_workflow PR.  That is a required temporary situation, until the regional_workflow PR is merged and the contributor updates his ufs-srweather-app PR with the hash of the authoritative regional_workflow develop branch.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: ... existing scripts (e.g., local variables should be lowercase, global variables should be uppercase).

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: The majority of the style used is based on standard shell scripting structure (like [this](https://google.github.io/styleguide/shellguide.html)).  I'll defer to @gsketefian regarding further comments on the current style, and whether we could refer to a specific online document.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: Contributors should update the .rst files in the same PR, or a subsequent PR, when possible.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: I'm wondering if we should call this "features" instead of "components".

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: Ideally, it would be all platforms, regardless of tier.  Unfortunately, it's unrealistic to test PRs on every platform from every tier.  At least until we get a fully-automated testing system set up, which is close.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: "When new capabilities are added or new bugs/issues are discovered, tests should be created and/or modified to test for these conditions." -> "When new capabilities are added or new bugs/issues are discovered, WE2E tests should be created and/or modified to verify new features are not broken in subsequent PRs."

**Review by JeffBeck-NOAA — CHANGES_REQUESTED** [review]

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: No-the #. is a numbering convention so that the list numbers are automatically populated. The comment above it will mess with formatting/spacing a bit until removed, but it's showing up just fine on RTD.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst:149]: Good idea! How do we go about that?

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: Fixed SRW App. As to PRs vs. PR's, it depends on which style guide we're using...

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: Isn't the regional_workflow its own fork? I guess this requirement is a little confusing to me, now that I think about it...

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: OK, yeah, I was seeing a different color of text below the comment, so I wanted to make sure.  Thanks.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: No, it was in the [Code Management document](https://docs.google.com/document/d/1KCGTjAG35H7tftVLBN6H96DGMnvL45LU5omvd2Ku24w/edit#heading=h.15mr0y6puhb9) under Coding Standards: General for the regional_workflow. It says: "All code must be indented appropriately, and keeping with the style of existing scripts."

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst:149]: In the .github/ISSUE_TEMPLATE directory of the repo, we'd just need to add a new "md" file for text.  Not too hard to do.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst:248]: Yes, but can anyone schedule a live code review? How would they do that if they're outside of NOAA? Would we be wondering who this random person is and ignore their request? Or is this actually likely to happen?

Comment (JeffBeck-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst:248]: That's a good question.  We've yet to run into this issue, but we've also yet to have a live code review request either.  Anyone could request one, but it would likely be someone with whom we've worked, or who has contacted us before.  We don't have a process outlined on how a live code review would work, but simply requesting one in the PR description would be a good first step.  Then we could work out a time, etc.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: OK, I think that was just to remind developers to stick to the established code style in the App.  If we want to refer to specific style guidelines, I'd defer to others on which ones.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: I think that's a good idea. "New components" could be confusing, since there are many UFS components that each have their own repos--we're not talking about those.

Comment (JeffBeck-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: Yeah, in the context you mentioned, "components" are much larger than just a new feature PR.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: It seems like NOAA's comms people use AP style guide, which says no apostrophe after multiple number/letter combos for plurals. But plurals of a single letter (e.g., "She got straight A's this semester") do have an apostrophe. Go figure. I don't like this rule, but I'll live with it. ;)

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: So, regional_workflow is it's own authoritative repository, not a fork.  The requirement to point to a developers fork in the Externals.cfg file is only valid if there is a corresponding PR for regional_workflow that goes hand-in-hand with the ufs-srweather-app.  This is only necessary when a change in ufs-srweather-app is required for the PR in regional_workflow to work (maybe a new CCPP suite is being introduced and supported in regional_workflow, for example).

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: Huh, I didn't even realize there was a choice here.  I thought the rule was that the apostrophe was for possessive use only.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst:149]: Ok, I gave it a shot! :) It's basically a copy of the feature request template, so feel free to suggest changes.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: Nope! Depends on the style guide. I originally learned MLA (way back), which used to require the apostrophe, but it seems like they changed that rule sometime in the past decade. The NYT [used to require](https://afterdeadline.blogs.nytimes.com/2010/04/13/faqs-on-style/) apostrophes after abbreviations that used periods (e.g., M.D.'s, Ph.D's), but idk if that's still the case. Seems like style guides are moving towards eliminating the apostrophe for plural abbreviations though. Good thing you made me do a little research on the matter, or I could've been stuck in the Stone Age forever. ;)

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: Oh, I just meant that for there to be a corresponding regional_workflow PR, the developer would have had to fork the regional_workflow repository. So he would have an SRW App fork and a regional_workflow fork, since it's not possible to have a single fork containing two repos (to my knowledge). Anyway, I'll try to clarify things in the guide, and you can tell me if my update is wrong. ;)

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: @JeffBeck-NOAA @mkavulich I think this is a _future_ feature in the Code Management Guide. Should I remove it until we can implement it?

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [inline docs/UsersGuide/source/ContributorsGuide.rst]: Here and elsewhere, I have written sentences such as "If a developer is working on a change, **they** should..." where the subject is a single developer, but the pronoun "they" referring back to the developer is technically plural. By prescriptive/traditional grammar standards, "they" should be replace by "he" or by "he or she." However, "they" is gender-neutral/nonbinary, and it seems like it is being chosen more frequently for this reason. Other ways to get around this problem involve wording things in the second person "you" or using passive voice, neither of which is ideal. Basically, there is no one right answer here, and I'm wondering what approach I should take. Do you all think the "they" approach works, or should we choose a different standard? @JeffBeck-NOAA

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: @gspetro-NOAA, while there is already Python code in the repo, the lint feature is not yet implemented.  I'll defer to @mkavulich for more details.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (JeffBeck-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: I think we can get around this problem by modifying the text to read, "Developers should X and Y".  Then the "they" would be appropriate, since "developers" is plural.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (mkavulich) [reply docs/UsersGuide/source/ContributorsGuide.rst]: @gspetro-NOAA We should take this out for now and leave it as a future addition, we have not yet set up any linting or a .pylintrc configuration file. Hopefully after this big round of PRs converting regional_workflow scripts to python we can start to work on this.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (jwolff-ncar) [inline docs/UsersGuide/source/ContributorsGuide.rst]: Super minor - The following steps should be "conducted" (or something along those lines rather than using followed again?)

**Review by JeffBeck-NOAA — APPROVED** [review]

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (jwolff-ncar) [inline docs/UsersGuide/source/ContributorsGuide.rst:241]: I think making a note to add labels in general would be really good. This could be a separate bullet but wanted to comment on it here since you mentioned adding high priority in this bullet.

Comment (jwolff-ncar) [reply docs/UsersGuide/source/ContributorsGuide.rst:241]: Never mind - see this below!

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (jwolff-ncar) [reply docs/UsersGuide/source/ContributorsGuide.rst:248]: It could also be raised the standing SRW code management review meetings on Thrs.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (jwolff-ncar) [inline docs/UsersGuide/source/ContributorsGuide.rst]: should this be environment.yaml?

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (jwolff-ncar) [inline docs/UsersGuide/source/ContributorsGuide.rst:332]: Will the document that @mkavulich created ultimately be linked in this chapter (once it exists on the wiki or elsewhere) for more details on this? Or is it all here, just reordered? I didn't do a thorough comparison.

**Review by jwolff-ncar — APPROVED** [review]
Minor comments/edits.

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: Changed to "The steps below should be followed..."

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst]: It's copied from @mkavulich 's code management doc. I've seen both. A brief internet search suggests that they are synonymous, but now that 3-letter file extensions are no longer a requirement on most systems, .yaml is more common. I'll change it to .yaml unless I hear (soon) from anyone that there's some reason to keep .yml

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gspetro-NOAA) [reply docs/UsersGuide/source/ContributorsGuide.rst:332]: I can definitely do that in an update if the SRW team wants that. I think everything that is pertinent to contributors is included here. A lot of the details from the code management doc apply only to the code managers, not to contributors (unless the contributors are also code managers, but then they probably have access to the doc anyway. ;) ). In other words, I view Mike's doc as the equivalent of a Teacher's Guide, whereas this UG chapter is more like "student instructions." The audience is a bit different. That said, couldn't hurt to include a link once the doc's ready!

File: docs/UsersGuide/source/ContributorsGuide.rst
Comment (gsketefian) [inline docs/UsersGuide/source/ContributorsGuide.rst]: To be clear about what "not login-enabled" means, add at the end `(i.e. do not use the -l flag).`

**Pull request merged**
